Sources/LanguageServerProtocolTransport/JSONRPCConnection.swift

```diff
@@ -295,45 +229,23 @@ public final class JSONRPCConnection: Connection {
       state = .running
       self.receiveHandler = receiveHandler
       self.closeHandler = closeHandler
+    }
 
-      receiveIO.read(offset: 0, length: Int.max, queue: queue) { done, data, errorCode in
-        guard errorCode == 0 else {
-          #if !os(Windows)
-          if errorCode != POSIXError.ECANCELED.rawValue {
-            logger.fault("IO error reading \(errorCode)")
-          }
-          #endif
-          if done { self.closeAssumingOnQueue() }
-          return
-        }
-
-        if done {
-          self.closeAssumingOnQueue()
-          return
-        }
-
-        guard let data = data, !data.isEmpty else {
-          return
-        }
+    // `readabilityHandler` is only ever called sequentially and all accesses to `parser` happen within its callback
+    // synchronously.
+    nonisolated(unsafe) let parser = JSONMessageParser(decoder: decodeJSONRPCMessage)
+    self.receiveFD.readabilityHandler = { fileHandle in
+      let data = orLog("Reading from \(self.name)") { try fileHandle.read(upToCount: parser.nextReadLength) }
+      guard let data, !data.isEmpty else {
+        // We have reached the end of `receiveFD`, close the connection. This will also set the `readabilityHandler` of
+        // `receiveFD` to `nil`, breaking the retain cycle.
+        self.close()
+        return
+      }
 
-        orLog("Writing input mirror file") {
-          try self.inputMirrorFile?.write(contentsOf: data)
-        }
-
-        // Parse and handle any messages in `buffer + data`, leaving any remaining unparsed bytes in `buffer`.
-        if self.requestBuffer.isEmpty {
-          data.withUnsafeBytes { (pointer: UnsafePointer<UInt8>) in
-            let rest = self.parseAndHandleMessages(from: UnsafeBufferPointer(start: pointer, count: data.count))
-            self.requestBuffer.append(contentsOf: rest)
-          }
-        } else {
-          self.requestBuffer.append(contentsOf: data)
-          var unused = 0
-          self.requestBuffer.withUnsafeBufferPointer { buffer in
-            let rest = self.parseAndHandleMessages(from: buffer)
-            unused = rest.count
-          }
-          self.requestBuffer.removeFirst(self.requestBuffer.count - unused)
+      self.queue.sync {
```

Comment:
Do we need the queue here? AFAIK `readabilityHandler` won’t be called multiple times concurrently – and if it was, then we could have races earlier on already.